Mathlib/RingTheory/PowerSeries/WeierstrassPreparation.lean

```diff
@@ -435,4 +454,191 @@ theorem weierstrassMod_zero_left [IsAdicComplete (IsLocalRing.maximalIdeal A) A]
+    · rw [(Polynomial.monic_X_pow n).leadingCoeff, H.isDistinguishedAt.monic.leadingCoeff]
+  · simp_rw [H.eq_mul, mul_assoc, IsUnit.mul_val_inv, mul_one, Polynomial.coe_sub,
+      Polynomial.coe_pow, Polynomial.coe_X, add_sub_cancel]
+
+section IsAdicComplete
+
+variable [IsAdicComplete (IsLocalRing.maximalIdeal A) A] {g : A⟦X⟧}
+  (hg : g.map (IsLocalRing.residue A) ≠ 0)
+include hg
+
+/-- **Weierstrass preparation theorem** ([washington_cyclotomic], Theorem 7.3):
+let `g` be a power series over a complete local ring,
+such that the image of `g` in the residue field is not zero. Then there exists a distinguished
+polynomial `f` and a power series `h` which is a unit, such that `g = f * h`. -/
+theorem exists_isWeierstrassFactorization : ∃ f h, g.IsWeierstrassFactorization f h := by
+  obtain ⟨q, r, H⟩ :=
+    (X ^ (g.map (IsLocalRing.residue A)).order.toNat).exists_isWeierstrassDivision hg
+  exact ⟨_, _, H.isWeierstrassFactorization hg⟩
+
+/-- The `f` in Weierstrass preparation theorem. -/
+noncomputable def weierstrassDistinguished : A[X] :=
+  (g.exists_isWeierstrassFactorization hg).choose
+
+/-- The `h` in Weierstrass preparation theorem. -/
+noncomputable def weierstrassUnit : A⟦X⟧ :=
+  (g.exists_isWeierstrassFactorization hg).choose_spec.choose
+
+theorem isWeierstrassFactorization_weierstrassDistinguished_weierstrassUnit :
+    g.IsWeierstrassFactorization (g.weierstrassDistinguished hg) (g.weierstrassUnit hg) :=
+  (g.exists_isWeierstrassFactorization hg).choose_spec.choose_spec
+
+theorem isDistinguishedAt_weierstrassDistinguished :
+    (g.weierstrassDistinguished hg).IsDistinguishedAt (IsLocalRing.maximalIdeal A) :=
+  (g.exists_isWeierstrassFactorization hg).choose_spec.choose_spec.isDistinguishedAt
+
+theorem isUnit_weierstrassUnit :
+    IsUnit (g.weierstrassUnit hg) :=
+  (g.exists_isWeierstrassFactorization hg).choose_spec.choose_spec.isUnit
+
+theorem eq_weierstrassDistinguished_mul_weierstrassUnit :
+    g = g.weierstrassDistinguished hg * g.weierstrassUnit hg :=
+  (g.exists_isWeierstrassFactorization hg).choose_spec.choose_spec.eq_mul
+
+end IsAdicComplete
+
+/-- The `f` and `h` in Weierstrass preparation theorem are unique.
+
+This result is stated using two `PowerSeries.IsWeierstrassFactorization` assertions, and only
+requires the ring being Hausdorff with respect to the maximal ideal. If you want `f` and `h` equal
+to `PowerSeries.weierstrassDistinguished` and `PowerSeries.weierstrassUnit`,
+use `PowerSeries.IsWeierstrassFactorization.unique` instead, which requires the ring being
+complete with respect to the maximal ideal. -/
+theorem IsWeierstrassFactorization.elim [IsHausdorff (IsLocalRing.maximalIdeal A) A]
+    {g : A⟦X⟧} {f f' : A[X]} {h h' : A⟦X⟧} (H : g.IsWeierstrassFactorization f h)
+    (H2 : g.IsWeierstrassFactorization f' h') : f = f' ∧ h = h' := by
+  obtain ⟨h1, h2⟩ := H.isWeierstrassDivision.elim H.map_ne_zero H2.isWeierstrassDivision
+  rw [← Units.ext_iff, inv_inj, Units.ext_iff] at h1
```

Comment:
Calling `ext_iff`  twice, once in reverse order and then in direct one, seems strange. Is there some API missing, or can this `rw` be simplified?

Reply:
Let me try.

Reply:
I can't find it; it should be the converse of [Units.inv_unique](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Group/Units/Basic.html#Units.inv_unique)...

Reply:
No problem